target/mips/helper.c

```diff
@@ -1486,6 +1486,48 @@ int cheri_tag_get(CPUMIPSState *env, target_ulong vaddr, int reg,
         return tagblk[CAP_TAGBLK_IDX(tag)];
 }
 
+int cheri_tag_get_many(CPUMIPSState *env, target_ulong vaddr, int reg,
+        hwaddr *ret_paddr)
+{
+    hwaddr paddr;
+    ram_addr_t ram_addr;
+    uint64_t tag;
+    uint8_t *tagblk;
+
+    qemu_log_mask(CPU_LOG_INSTR, "%s: va=%lx\n", __func__, vaddr);
+
+    /* XXX MMU_DATA_CAP_LOAD?  Causes some bizarre misbehavior on CheriBSD
+     * if so, tho'.  Needs further debugging.
+     */
+    paddr = v2p_addr(env, vaddr, MMU_DATA_LOAD, reg);
+
+    if (ret_paddr)
+        *ret_paddr = paddr;
+
+    ram_addr = p2r_addr(env, paddr, NULL);
+    if (ram_addr == -1LL)
+        return 0;
+
+    /* Get the base tag number and tag block ptr */
+    tag = (ram_addr >> (CAP_TAG_SHFT + 3)) << 3; // XXX 2^3 capabilities per cache line
```

Comment:
This probably needs an ifdef for 128 vs 256

Reply:
Yeah, sensible.  Is the cacheline width specified elsewhere in QEMU?